library/src/main/java/com/feedhenry/sdk/android/ClientIdGeneratorImpl.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright Red Hat, Inc, and individual contributors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.feedhenry.sdk.android;
+
+import android.annotation.SuppressLint;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.provider.Settings;
+import com.feedhenry.sdk.utils.ClientIdGenerator;
+
+/**
+ * Generates unique client id on Android.
+ */
+public class ClientIdGeneratorImpl implements ClientIdGenerator {
+
+    private static final String FH_SYNC_PREFERENCES = "fh-sync";
+    private static final String KEY_CLIENT_ID = "clientId";
+
+    private final Context context;
+
+    public ClientIdGeneratorImpl(Context ctx) {
+        context = ctx;
+    }
+
+    @Override
+    public String getClientId() {
+        SharedPreferences prefs = context.getSharedPreferences(FH_SYNC_PREFERENCES, Context.MODE_PRIVATE);
+        String clientId = prefs.getString(KEY_CLIENT_ID, generateNewId());
+        prefs.edit().putString(KEY_CLIENT_ID, clientId).apply();
```

Comment:
this will write the clientId *everytime* when `getClientId` is invoked. Is this the desired behaviour?

Reply:
Maybe not exactly, it should do that only when it wasn't previously defined. However the implementation writes only changes but not the same value.
https://github.com/android/platform_frameworks_base/blob/master/core/java/android/app/SharedPreferencesImpl.java#L496-L501
Maybe I shouldn't rely on hidden implementation.

Reply:
good to know that, but I think it's better to only commit the changes if the value is not defined in our code to make it clear and obvious.